### PR TITLE
Normalize building kind and add kind_detail

### PR DESCRIPTION
**Pull request by rmarianski**
Connects to https://github.com/tilezen/vector-datasource/issues/842

@nvkelso, @zerebubuth could you review please?

### CODE REVIEW

**Comment by nvkelso** [comment]
LGTM

File: data/functions.sql [hunk 1]
```diff
@@ -687,3 +687,14 @@ BEGIN
...
+RETURNS TEXT AS $$
+DECLARE
+BEGIN
+  RETURN CASE WHEN val = 'yes' THEN NULL
+              WHEN val IN ('house', 'residential', 'garage', 'apartments', 'hut', 'industrial', 'roof', 'shed', 'detached', 'commercial', 'garages', 'terrace', 'school', 'retail', 'farm_auxiliary', 'construction', 'church', 'barn', 'greenhouse', 'service', 'cabin', 'manufacture', 'farm', 'warehouse', 'civic', 'office', 'collapsed', 'university', 'public', 'hangar', 'entrance', 'hospital', 'static_caravan', 'hotel', 'chapel', 'semidetached_house', 'damaged', 'kindergarten', 'transportation', 'train_station', 'dormitory', 'storage_tank', 'trullo', 'stable', 'ger', 'houseboat', 'shop', 'college', 'ruins', 'semi', 'bunker', 'pajaru', 'mosque', 'bungalow', 'slurry_tank', 'supermarket', 'transformer_tower', 'duplex', 'factory', 'cowshed', 'constructie', 'dam', 'agricultural', 'tank', 'shelter', 'dwelling_house', 'temple', 'tower', 'agricultural', 'tank', 'shelter', 'dwelling_house', 'temple', 'tower', 'silo', 'residence', 'pavilion', 'store', 'support', 'allotment_house', 'summer_cottage', 'mobile_home', 'administrative', 'kiosk', 'other', 'boathouse', 'glasshouse', 'storage') THEN val
```
Comment (zerebubuth) [inline data/functions.sql]: I wonder if we should exclude the `building=collapsed` and `building=ruins` from inclusion as a `kind:building`, perhaps by excluding them from the definition of `*is_building` in the YAML?

And perhaps also `building=no`, which at the moment would produce `kind:building` but no kind detail, which could be a bit misleading.

Comment (zerebubuth) [inline data/functions.sql]: Previously, we would have given a `building=no` a `kind:no`, which might also have been unexpected and confusing.

File: yaml/buildings.yaml [hunk 2]
```diff
@@ -18,6 +18,12 @@ globals:
       roof_orientation: {col: "roof:orientation"}
       addr_housenumber: {col: "addr:housenumber"}
       addr_street: {col: "addr:street"}
+  - &building_kind_detail
+      kind_detail:
+        expr: mz_building_kind_detail(building)
+  - &building_part_kind_detail
+      kind_detail:
+        expr: mz_building_kind_detail("building:part")
```
Comment (zerebubuth) [inline yaml/buildings.yaml]: What's the overlap between the popular `building:part` values and `building` values? Are there any popular `building:part` values we're missing from the list in `mz_building_kind_detail()`?

File: data/functions.sql (hunk 1 above)
Comment (rmarianski) [inline data/functions.sql]: > And perhaps also building=no, which at the moment would produce kind:building but no kind detail, which could be a bit misleading.

Aren't we already [filtering those out](https://github.com/tilezen/vector-datasource/blob/45af11b772bdd1eec09d4285fbf00883d3334984/yaml/buildings.yaml#L21-L28)? 

> I wonder if we should exclude the building=collapsed and building=ruins from inclusion

Sounds good to me, or even treat them as we do [closed names](https://github.com/tilezen/vector-datasource/issues/483), ie only show them at z17+. @nvkelso what do you think?

File: yaml/buildings.yaml (hunk 2 above)
Comment (rmarianski) [inline yaml/buildings.yaml]: Good point, I hadn't considered the different popular values.

Ignoring misspellings, capitalizations, numbers, "y" and "true", here's what's in `building:part` that isn't in `building` in the first 4 pages according to taginfo:

```
apartment
arch
balcony
barracks
base
basement
bell_tower
bridge
bus_station
castle
chimney
church:part
column
corridor
Corridor
default
dome
door
elevator
floor
hall
historic
horizontal
main
museum
offices
parking
part
passageway
pillar
porch
proposed
ramp
room
silos
stairs
stairway
steps
stilobate
technical
tier
vertical
verticalpassage
wall
window
```

Should we just have separate `kind_detail` whitelists for `building` and `building:part`?

File: data/functions.sql (hunk 1 above)
Comment (zerebubuth) [inline data/functions.sql]: > > And perhaps also building=no, which at the moment would produce kind:building but no kind detail, which could be a bit misleading.
>
> Aren't we already filtering those out?

D'oh! You're absolutely right - I hadn't expanded the non-diff parts enough to read that, sorry!

File: yaml/buildings.yaml (hunk 2 above)
Comment (zerebubuth) [inline yaml/buildings.yaml]: Some of these (`dome`, `part`, `main`, `floor`) seem like they could be useful for rendering, although I don't know a huge amount about the 2.5D building/part stuff.

Values like `historic` and `proposed` make me a bit nervous - it sounds like they may be applied to buildings which aren't currently present in reality. Might we want to give it a different kind to help make it clearer to someone using the data that it's different and, for example, might overlap with other buildings or features which pre- or post-date the historic / proposed / collapsed / ruined -building?

Comment (zerebubuth) [inline yaml/buildings.yaml]: Nitpick - there's `Corridor` and `corridor` on the list.

Comment (rmarianski) [inline yaml/buildings.yaml]: Good catch. While diff'ing, I also realized that I had some dupes in the original building list. Updated in e59ecdc.

File: data/functions.sql (hunk 1 above)
Comment (nvkelso) [inline data/functions.sql]: > I wonder if we should exclude the building=collapsed and building=ruins from inclusion as a kind:building, perhaps by excluding them from the definition of *is_building in the YAML?

If a building is **collapsed** it still has a "footprint" _but we should probably zero out it's height_? I'm thinking of Christchurch earthquake in New Zealand and the HOT work around mapping all those buildings – we'd still want them to show up.

For buildings marked **ruins**, it looks like there is a tiny bit of overlap between [building=ruins](http://wiki.openstreetmap.org/wiki/Tag:building%3Druins) and [historic=ruin](http://wiki.openstreetmap.org/wiki/Tag:historic%3Druins). But they are different concepts. If it were historic I'd expect a POI to be generated, but for these basic buildings I think we'd only generate a label placement and pass thru `ruins` as the `kind_detail`. Ruins can still have height, so I wouldn't mess with that value.

I can see only showing the label position for these at zoom 17+, but the features themselves should be visible like normal based on their area.

File: yaml/buildings.yaml (hunk 2 above)
Comment (nvkelso) [inline yaml/buildings.yaml]: I expect our kind details to more closely match what's in our existing kind list for landuse, and maybe a few POI values. This also lines up with what's on the OSM wiki page: http://wiki.openstreetmap.org/wiki/Key:building.

Some of these are sounding more like `roof:shape` which there's regular (every month it seems) interest in adding to tiles though there is currently no Tangram support for it. http://wiki.openstreetmap.org/wiki/Simple_3D_buildings

`proposed` is interesting, we should do what we do for transit stations for that one.

Should building parts have an additional property to store some of these other random values?

Comment (bcamper) [inline yaml/buildings.yaml]: We already have `roof:shape` in tiles!

http://tangrams.github.io/explorer/#16.5/40.7104/-74.0123/roof_shape/dome

Comment (zerebubuth) [inline yaml/buildings.yaml]: Hmm.... does that mean someone added `building:part=dome` when they meant `roof:shape=dome`? If so, we should probably ignore the `building:part`. Looks like there's [22 instances](http://taginfo.openstreetmap.org/keys/building%3Apart#values) of `building:part=dome`, so it's likely a typo / mis-tag.

There's so few of these values (other than `yes` & `roof`) that perhaps I was wrong to suggest we include them in the list as well.

**Comment by zerebubuth** [comment]
Sorry - I think my previous comments about `building:part` values were a wild goose chase. Contrary to what I thought, looks like there's not much information there.

:+1: looks good as it is at the moment - no `building:part` changes needed.

**Comment by rmarianski** [comment]
@nvkelso is this ok to merge as is?

There were a couple of related thoughts here on how to update the behavior , but it seems like the tag scenarios appear so infrequently that it might not be worth it. Can we create follow up issues for any specifics that we want to follow up on?

**Comment by rmarianski** [comment]
@nvkelso can you review the list of [values whitelisted](https://github.com/tilezen/vector-datasource/pull/886/files#diff-85235d701f80586d926a4211261dbbc1R697) to appear in `kind_detail`? This list was the first four pages of [taginfo for building](http://taginfo.openstreetmap.org/keys/building#values).

**Comment by nvkelso** [comment]
Generally things 'thing' quoted where your list, unquoted are my additions. But remake using these values.

I'm not 100% sure about the residential stuff. There was also a bunch of Indonesian words that I tried to replace with British English values.

### BUILDING PARTS

Whitelist for `kind_building_part` (new property in addition to `kind` and `kind_details` tags):

**default values:**

_discuss!!!_

yes ('yes')
part ('yes')
church:part ('yes')
default ('yes')

**good unique values:**

balcony
base
column
elevator
entrance
floor
pillar
room
hall
main
passageway
porch
ramp
roof
steps
stilobate
tier
tower
verticalpassage
wall
window
door
arch

**Building part replacements:** 

corridor ('verticalpassage')
Corridor ('verticalpassage')
vertical ('verticalpassage')
vertical_passage ('verticalpassage')
stairs ('steps')
stairway ('steps'

### BUILDING_KIND

**these stand on their own (aren't a kind from landuse/pois)**

'collapsed'
'construction'
'damaged'

abandoned
proposed
destroyed
disused

'residential'
'administrative'
'agricultural'
'allotment_house'
'apartments'
'barn'
'boathouse'
'bungalow'
'bunker'
'cabin'
'chapel'
'church'
'civic'
'college'
'commercial'
'cowshed'
'dam'
'detached'
'dormitory'
'factory'
'farm'
'farm_auxiliary'
'garage'
'garages'
'ger'
'glasshouse'
'greenhouse'
'hangar'
'hospital'
'hotel'
'houseboat'
'hut'
'industrial'
'kindergarten'
'kiosk'
'manufacture'
'mobile_home'
'mosque'
'office'
'pavilion'
'public'
'retail'
'roof'
'ruin'
'ruins'
'school'
'service'
'shed'
'shelter'
'shop'
'silo'
'slurry_tank'
'stable'
'static_caravan'
'storage'
'storage_tank'
'store'
'summer_cottage'
'supermarket'
'tank'
'temple'
'terrace'
'tower'
'train_station'
'transformer_tower'
'transportation'
'university'
'warehouse'

grandstand
carport
veranda
pub
cathedral
wayside_shrine
outbuilding
castle
bridge
beach_hut
utility
government
monastery
gazebo
prison
stadium
container
manor
basilica
restaurant
museum
healthcare
depot
bell_tower
substation
shrine
clubhouse
hermitage
power
summer_house
synagogue
convent
clinic
library
mall
parking
arbour
airport
chimney
bank
mortuary
works
brewery
fire_station

**Building kind replacements:** 

'static caravan' ('static_caravan')
'storage tank' ('storage_tank')
railway_station ('train_station')
stables ('stable')
'perumahan permukiman' ('residential')
barne ('barn')
prefab_container ('container)
public_building ('public')
greenhouse_horticulture ('greenhouse')
semi_detached ('semidetached_house')
flat ('apartments')
houses ('residential')
roof=permanent ('roof')
offices ('office')
residentiel1 ('residential')
semi-detached ('semidetached_house')
residences ('residential')
'residence' ('residential')
apartment ('apartments')
station ('transportation')
'constructie' ('construction')
'semi' ('semidetached_house')
'dwelling_house' ('residential')
'duplex' ('residential')
'semidetached_house' ('residential')
education ('school')
townhome ('residential')
'house' ('residential')

'commercial;residential' ('mixed_use')  -- this is made up


**we'd ignore these:** 

'Rumah Masyarakat' ('yes')
rumah ('yes')
rumah_penduduk ('yes')
Rumah ('yes')
true ('yes')
building ('yes')
bangunan ('yes') - see https://en.wiktionary.org/wiki/bangunan
'trullo' ('yes')
'other' ('yes')

**Comment by nvkelso** [comment]
Complication, when we start showing building parts at zoom 15, some features are tagged as both `building=*` and `building:part=*`. At earlier zooms the "bottom" one should be a `kind:building` and the then at zoom 15+ we'd call it a `building_part`.

The area and volume calculations should be updated once #653 lands, too, so we don't clip of certain building parts zooming in and out.

**Comment by rmarianski** [comment]
Test failures should be resolved in b5f8cdd.

**Comment by nvkelso** [comment]
New version 💯, ready to merge!

**Pull request merged**